File: src/components/internal-browser/browser-events.ts
import {
  BROWSER_EVENT_VERSION,
  type BrowserSessionEvent,
  type BrowserSessionState,
  type BrowserTabEvent as BrowserTabEventPayload,
  type PopupRequestEvent,
} from "./browser-contract.generated";

export const BROWSER_SESSION_EVENT = "browser-tab-state-changed-v1";
export const BROWSER_POPUP_EVENT = "browser-popup-request-v1";
export const BROWSER_READY_EVENT = "browser-view-ready-v1";
export const BROWSER_ENGINE_STOPPED_EVENT = "browser-engine-stopped-v1";
export const BROWSER_BLOCKED_FEATURE_EVENT = "browser-feature-blocked-v1";
export { BROWSER_EVENT_VERSION } from "./browser-contract.generated";
export type { BrowserTabCreation } from "./browser-contract.generated";
export type BrowserPopupRequest = Omit<PopupRequestEvent, "eventVersion" | "conversationId">;
export type BrowserTabEvent = Omit<BrowserTabEventPayload, "eventVersion" | "conversationId">;

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}

export function parseSessionEvent(
  value: unknown,
  conversationId: string,
): BrowserSessionState | null {
  if (
    !isRecord(value) || value.eventVersion !== BROWSER_EVENT_VERSION ||
    value.conversationId !== conversationId
  ) return null;
  return (value as BrowserSessionEvent).session;
}

export function parsePopupEvent(
  value: unknown,
  conversationId: string,
): BrowserPopupRequest | null {
  if (
    !isRecord(value) || value.eventVersion !== BROWSER_EVENT_VERSION ||
    value.conversationId !== conversationId
  ) return null;
  const { generation, sourceTabId, url } = value as PopupRequestEvent;
  return { generation, sourceTabId, url };
}

export function parseBrowserTabEvent(
  value: unknown,
  conversationId: string,
): BrowserTabEvent | null {
  if (
    !isRecord(value) || value.eventVersion !== BROWSER_EVENT_VERSION ||
    value.conversationId !== conversationId
  ) return null;
  const { generation, tabId } = value as BrowserTabEventPayload;
  return { generation, tabId };
}

File: src-tauri/src/services/browser/browser_events.rs
use serde::Serialize;
#[cfg(native_browser)]
use std::sync::atomic::{AtomicU64, Ordering};

#[cfg(native_browser)]
pub(super) const VIEW_READY_EVENT: &str = "browser-view-ready-v1";
#[cfg(native_browser)]
pub(super) const POPUP_REQUEST_EVENT: &str = "browser-popup-request-v1";
#[cfg(native_browser)]
pub(super) const ENGINE_STOPPED_EVENT: &str = "browser-engine-stopped-v1";
#[cfg(native_browser)]
pub(super) const BLOCKED_FEATURE_EVENT: &str = "browser-feature-blocked-v1";
#[cfg(native_browser)]
pub(super) const TAB_STATE_CHANGED_EVENT: &str = "browser-tab-state-changed-v1";
pub(super) const EVENT_VERSION: u8 = 1;

#[cfg(native_browser)]
static EVENT_GENERATION: AtomicU64 = AtomicU64::new(1);

#[cfg(native_browser)]
pub(super) fn next_event_generation() -> Option<u64> {
    EVENT_GENERATION
        .fetch_update(Ordering::AcqRel, Ordering::Acquire, |current| {
            current.checked_add(1)
        })
        .ok()
}

#[derive(Clone, Serialize)]
#[cfg_attr(test, derive(ts_rs::TS))]
#[serde(rename_all = "camelCase")]
pub(super) struct BrowserTabEvent {
    pub event_version: u8,
    #[cfg_attr(test, ts(type = "number"))]
    pub generation: u64,
    pub conversation_id: String,
    pub tab_id: String,
}

#[derive(Clone, Serialize)]
#[cfg_attr(test, derive(ts_rs::TS))]
#[serde(rename_all = "camelCase")]
pub(super) struct PopupRequestEvent {
    pub event_version: u8,
    #[cfg_attr(test, ts(type = "number"))]
    pub generation: u64,
    pub conversation_id: String,
    pub source_tab_id: String,
    pub url: String,
}

#[derive(Clone, Serialize)]
#[cfg_attr(test, derive(ts_rs::TS))]
#[serde(rename_all = "camelCase")]
pub(super) struct BrowserSessionEvent {
    pub event_version: u8,
    pub conversation_id: String,
    pub session: super::session_model::BrowserSessionState,
}
